src/Chainweb/Pact/Backend/RelationalCheckpointer.hs

```diff
@@ -296,19 +296,39 @@ doRegisterSuccessful :: Db -> PactHash -> IO ()
 doRegisterSuccessful dbenv (TypedHash hash) =
     runBlockEnv dbenv (indexPactTransaction $ BS.fromShort hash)
 
+doLookupSuccessful :: Db -> Maybe ConfirmationDepth -> V.Vector PactHash -> IO (HashMap.HashMap PactHash (T2 BlockHeight BlockHash))
+doLookupSuccessful dbenv confDepth hashes = runBlockEnv dbenv $ do
+    r <- callDb "doLookupSuccessful" $ \db -> do
+      let
+        currentHeightQ = "SELECT blockheight FROM BlockHistory \
+            \ ORDER BY blockheight DESC LIMIT 1"
 
-doLookupSuccessful :: Db -> PactHash -> IO (Maybe (T2 BlockHeight BlockHash))
-doLookupSuccessful dbenv (TypedHash hash) = runBlockEnv dbenv $ do
-    r <- callDb "doLookupSuccessful" $ \db ->
-         qry db qtext [ SBlob (BS.fromShort hash) ] [RInt, RBlob] >>= mapM go
-    case r of
-        [] -> return Nothing
-        (!o:_) -> return (Just o)
+      -- if there is a confirmation depth, we get the current height and calculate
+      -- the block height, to look for the transactions in range [0, current block height - confirmation depth]
+      blockheight <- case confDepth of
+        Nothing -> pure Nothing
+        Just (ConfirmationDepth cd) -> do
+          currentHeight <- qry_ db currentHeightQ [RInt]
+          case currentHeight of
+            [[SInt bh]] -> pure $ Just (bh - fromIntegral cd)
+            _ -> fail "impossible"
+
+      let
+        blockheightval = maybe [] (\bh -> [SInt bh]) blockheight
+        qvals = [ SBlob (BS.fromShort hash) | (TypedHash hash) <- V.toList hashes ] ++ blockheightval
+        qtypes = RInt : [ RBlob | _ <- V.toList hashes ] ++ maybe [] (const [RInt]) blockheight
+
+      qry db qtext qvals qtypes >>= mapM go
+    return $ HashMap.fromList (zip (V.toList hashes) r)
   where
     qtext = "SELECT blockheight, hash FROM \
             \TransactionIndex INNER JOIN BlockHistory \
-            \USING (blockheight) WHERE txhash = ?;"
-    go [SInt h, SBlob blob] = do
+            \USING (blockheight) WHERE txhash IN (" <> hashesParams <> ")"
+            <> maybe "" (const " AND blockheight <= ?") confDepth
+            <> ";"
+    hashesParams = Utf8 $ intercalate "," [ "?" | _ <- V.toList hashes]
+
+    go ((SInt h):(SBlob blob):_) = do
```

Comment:
The only suspicious thing of this PR to me.

I don't know why but sometimes the results that are being returned include not two element lists, but three elements and the last one is an empty blob `SBlob ""`.

Reply:
This is probably due to the above `qtypes` business.